# **San Francisco city employee salary data Analysis:**

##### **Data Source:** 
> __https://www.kaggle.com/datasets/kaggle/sf-salaries?select=Salaries.csv__

This data contains the `names`, `job title`, and `compensation` for San Francisco city employees on an annual basis from `2011 to 2014`.

#### **Exploration Ideas:**
To help get us started, here are some data exploration ideas:

- How have salaries changed over time between different groups of people?

- How are base pay, overtime pay, and benefits allocated between different groups? 

- Is there any evidence of pay discrimination based on gender in this dataset?

- How is budget allocated based on different groups and responsibilities?

In [1]:
import pandas as pd 
import numpy as np 

In [2]:
# Impoorting dataset:
data= pd.read_csv("C:/Users/MyMachine/Desktop/Mission-Project/00_DataSets/06_San Francisco city employee salary data.csv", low_memory=False)

In [3]:
data.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

In [4]:
data.columns

Index(['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')

In [5]:
len(data.columns) #Total number of columns

13

In [6]:
len(data) # Total number of observations

148654

In [7]:
data.dtypes # data type of each column

Id                    int64
EmployeeName         object
JobTitle             object
BasePay              object
OvertimePay          object
OtherPay             object
Benefits             object
TotalPay            float64
TotalPayBenefits    float64
Year                  int64
Notes               float64
Agency               object
Status               object
dtype: object

In [8]:
data.sample() # a random observation from the dataset

Id    EmployeeName       JobTitle   BasePay OvertimePay OtherPay  \
140653  140654  Cynthia R Mott  Special Nurse  19360.88     2076.22  3097.28   

       Benefits  TotalPay  TotalPayBenefits  Year  Notes         Agency Status  
140653  4480.53  24534.38          29014.91  2014    NaN  San Francisco     PT

In [9]:
data.shape # The shape of the dataframe

(148654, 13)

In [10]:
# Total number of rows: 
len(data)

148654

In [11]:
# Total number of columns: 
len(data.columns)

13

In [12]:
# The memory requrement:

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148049 non-null  object 
 4   OvertimePay       148654 non-null  object 
 5   OtherPay          148654 non-null  object 
 6   Benefits          112495 non-null  object 
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(3), int64(2), object(8)
memory usage: 14.7+ MB


In [13]:
#  Check null values in the dataset: 
data.isnull().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Some columns like ID, Notes, Agency and Status columns are not necessary for our canse. So, we will drop the unusable columns.

In [14]:
data.columns

Index(['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')

In [15]:
# Drop unusable columns: 

data.drop(["Id", "Notes", "Status", "Agency"], inplace= True, axis= 1)

In [16]:
data.shape

(148654, 9)

In [17]:
# Get overall statistics of the dataset:
data.describe(include= 'all')

EmployeeName          JobTitle BasePay OvertimePay OtherPay Benefits  \
count        148654            148654  148049      148654   148654   112495   
unique       110811              2159  109724       66162    84243    98647   
top       Kevin Lee  Transit Operator     0.0         0.0      0.0      0.0   
freq             13              7036     872       57868    30068     1024   
mean            NaN               NaN     NaN         NaN      NaN      NaN   
std             NaN               NaN     NaN         NaN      NaN      NaN   
min             NaN               NaN     NaN         NaN      NaN      NaN   
25%             NaN               NaN     NaN         NaN      NaN      NaN   
50%             NaN               NaN     NaN         NaN      NaN      NaN   
75%             NaN               NaN     NaN         NaN      NaN      NaN   
max             NaN               NaN     NaN         NaN      NaN      NaN   

             TotalPay  TotalPayBenefits           Year  
count   148654.000000     148654.000000  148654.000000  
unique            NaN               NaN            NaN  
top               NaN               NaN            NaN  
freq              NaN               NaN            NaN  
mean     74768.321972      93692.554811    2012.522643  
std      50517.005274      62793.533483       1.117538  
min       -618.130000       -618.130000    2011.000000  
25%      36168.995000      44065.650000    2012.000000  
50%      71426.610000      92404.090000    2013.000000  
75%     105839.135000     132876.450000    2014.000000  
max     567595.430000     567595.430000    2014.000000

In [18]:
# find occurrences of the employeee names: top 10:
data.columns
data["EmployeeName"].value_counts().head(10)

EmployeeName
Kevin Lee       13
William Wong    11
Richard Lee     11
Steven Lee      11
John Chan        9
Michael Wong     9
KEVIN LEE        9
Stanley Lee      9
William Lee      8
MICHAEL LEE      8
Name: count, dtype: int64

In [19]:
# Find the number of unique job titles:
data.columns
data["JobTitle"].nunique()

2159

In [20]:
# Total number of job titles that contain "Captain" word:
data["JobTitle"].apply(lambda x: "Captain" in x).value_counts()
# Only one condition can be written with lambda function
# but here in actual, we have to include two conditions:
# 1. Job title contains "Captain" word and 
# 2. Job title is case insensitive

# Next process for this is more suitable and should follow:

JobTitle
False    148243
True        411
Name: count, dtype: int64

In [21]:
data["JobTitle"].str.contains("Captain", case= False).value_counts()
# As the case insensitivity is appilied, the number of observations also increase from 411 to 552

JobTitle
False    148102
True        552
Name: count, dtype: int64

In [22]:
# Display all the employee names from fire department:
(data["JobTitle"]== "Fire").value_counts()

JobTitle
False    148654
Name: count, dtype: int64

In [23]:
data.shape

(148654, 9)

So, there is no separate department that is specific: "Fire". "Fire" might be in combination with other relevent words which is what we will explore next:

In [24]:
data["JobTitle"].str.contains("Fire", case= False).value_counts()

JobTitle
False    142775
True       5879
Name: count, dtype: int64

So, the 5879 employees are from the fore department or are related to the fire department.

In [25]:
data[data["JobTitle"].str.contains("Fire", case= False)]["EmployeeName"]

4            PATRICK GARDNER
6                  ALSON LEE
8             MICHAEL MORRIS
9         JOANNE HAYES-WHITE
10             ARTHUR KENNEY
                 ...        
145956      Kenneth C Farris
147556         Edward A Dunn
148021        Kari A Johnson
148209          Sheryl K Lee
148554       Lawrence F Gatt
Name: EmployeeName, Length: 5879, dtype: object

In [26]:
# Find minimum, maximum and average BasePay:
data["BasePay"].replace("Not Provided", np.nan).astype(dtype=float).describe()

count    148045.000000
mean      66325.448840
std       42764.635495
min        -166.010000
25%       33588.200000
50%       65007.450000
75%       94691.050000
max      319275.010000
Name: BasePay, dtype: float64

In [27]:
# First replace "Not Provided" with NaN, then convert to float
data["BasePay"] = data["BasePay"].replace("Not Provided", np.nan).astype(float)

# Now we can get the statistics
print(f"Minimum: {data['BasePay'].min()}")
print(f"Maximum: {data['BasePay'].max()}")
print(f"Average: {data['BasePay'].mean()}")

Minimum: -166.01
Maximum: 319275.01
Average: 66325.4488404877


In [28]:
# Replace "Not Provided" in EmployeeName Column to NaN:
data["EmployeeName"].replace("Not provided", np.nan)

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650                  NaN
148651                  NaN
148652                  NaN
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

In [29]:
data

EmployeeName                                        JobTitle  \
0          NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1            GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2          ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3       CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4         PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
...                   ...                                             ...   
148649      Roy I Tillery                                       Custodian   
148650       Not provided                                    Not provided   
148651       Not provided                                    Not provided   
148652       Not provided                                    Not provided   
148653          Joe Lopez                      Counselor, Log Cabin Ranch   

          BasePay   OvertimePay      OtherPay      Benefits   TotalPay  \
0       167411.18           0.0     400184.25           NaN  567595.43   
1       155966.02     245131.88     137811.38           NaN  538909.28   
2       212739.13     106088.18       16452.6           NaN  335279.91   
3        77916.00      56120.71      198306.9           NaN  332343.61   
4       134401.60        9737.0     182234.59           NaN  326373.19   
...           ...           ...           ...           ...        ...   
148649       0.00          0.00          0.00          0.00       0.00   
148650        NaN  Not Provided  Not Provided  Not Provided       0.00   
148651        NaN  Not Provided  Not Provided  Not Provided       0.00   
148652        NaN  Not Provided  Not Provided  Not Provided       0.00   
148653       0.00          0.00       -618.13          0.00    -618.13   

        TotalPayBenefits  Year  
0              567595.43  2011  
1              538909.28  2011  
2              335279.91  2011  
3              332343.61  2011  
4              326373.19  2011  
...                  ...   ...  
148649              0.00  2014  
148650              0.00  2014  
148651              0.00  2014  
148652              0.00  2014  
148653           -618.13  2014  

[148654 rows x 9 columns]

In [30]:
data.replace("Not provided", np.nan, inplace=True)
data.replace("Not Provided", np.nan, inplace=True)

In [31]:
data.isnull().sum()

EmployeeName            6
JobTitle                4
BasePay               609
OvertimePay             4
OtherPay                4
Benefits            36163
TotalPay                0
TotalPayBenefits        0
Year                    0
dtype: int64

In [32]:
data.shape

(148654, 9)

In [33]:
# Drop the rows having more than 5 missing values:
data.drop(data[(data.isna().sum(axis=1)>=4)].index, axis= 0, inplace= True)

In [34]:
data

EmployeeName                                        JobTitle  \
0          NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1            GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2          ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3       CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4         PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
...                   ...                                             ...   
148645   Carolyn A Wilson                       Human Services Technician   
148647     Joann Anderson                     Communications Dispatcher 2   
148648        Leon Walker                                       Custodian   
148649      Roy I Tillery                                       Custodian   
148653          Joe Lopez                      Counselor, Log Cabin Ranch   

          BasePay OvertimePay   OtherPay Benefits   TotalPay  \
0       167411.18         0.0  400184.25      NaN  567595.43   
1       155966.02   245131.88  137811.38      NaN  538909.28   
2       212739.13   106088.18    16452.6      NaN  335279.91   
3        77916.00    56120.71   198306.9      NaN  332343.61   
4       134401.60      9737.0  182234.59      NaN  326373.19   
...           ...         ...        ...      ...        ...   
148645       0.00        0.00       0.00     0.00       0.00   
148647       0.00        0.00       0.00     0.00       0.00   
148648       0.00        0.00       0.00     0.00       0.00   
148649       0.00        0.00       0.00     0.00       0.00   
148653       0.00        0.00    -618.13     0.00    -618.13   

        TotalPayBenefits  Year  
0              567595.43  2011  
1              538909.28  2011  
2              335279.91  2011  
3              332343.61  2011  
4              326373.19  2011  
...                  ...   ...  
148645              0.00  2014  
148647              0.00  2014  
148648              0.00  2014  
148649              0.00  2014  
148653           -618.13  2014  

[148650 rows x 9 columns]

In [35]:
# To find job title of ALLBERT PARDINI: 
data.columns
data[data["EmployeeName"].str.contains("ALBERT PARDINI", case= False, na= False)]["JobTitle"]

2        CAPTAIN III (POLICE DEPARTMENT)
36519                          Captain 3
Name: JobTitle, dtype: object

In [36]:
# ANother approach: 
data[data["EmployeeName"]== "ALBERT PARDINI"]["JobTitle"]

2    CAPTAIN III (POLICE DEPARTMENT)
Name: JobTitle, dtype: object

In [37]:
# How much ALBERT PARDINI make (include Benefits):
data.columns
data[data["EmployeeName"]== "ALBERT PARDINI"]["TotalPayBenefits"]

2    335279.91
Name: TotalPayBenefits, dtype: float64

In [38]:
# Display the name of the person having the higest BasePay:
data.columns
data["BasePay"].max()

np.float64(319275.01)

In [39]:
data[data["BasePay"].max() == data["BasePay"]]["EmployeeName"]

72925    Gregory P Suhr
Name: EmployeeName, dtype: object

In [40]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [41]:
data['Year'].value_counts()

Year
2014    38119
2013    37606
2012    36766
2011    36159
Name: count, dtype: int64

In [42]:
# Find the average basepay of all employee per year:
data.groupby("Year")["BasePay"].mean()

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

In [43]:
# Find the average basepay of all employee per job title:
data.groupby("JobTitle")["BasePay"].mean()

JobTitle
ACCOUNT CLERK                                     43300.806506
ACCOUNTANT                                        46643.172000
ACCOUNTANT INTERN                                 28732.663958
ACPO,JuvP, Juv Prob (SFERS)                       62290.780000
ACUPUNCTURIST                                     66374.400000
                                                      ...     
X-RAY LABORATORY AIDE                             47664.773077
X-Ray Laboratory Aide                             46086.387100
YOUTH COMMISSION ADVISOR, BOARD OF SUPERVISORS    52609.910000
Youth Comm Advisor                                39077.957500
ZOO CURATOR                                       43148.000000
Name: BasePay, Length: 2158, dtype: float64

In [44]:
# Find the average basepay of employee having job title "ACCOUNTANT":
data.groupby("JobTitle")["BasePay"].mean()["ACCOUNTANT"]

np.float64(46643.172)

In [45]:
# Find the average base Pay of employee having job title "ACCOUNTANT":
data[data["JobTitle"]== "ACCOUNTANT"]["BasePay"].mean()

np.float64(46643.172)

In [47]:
data.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [50]:
# Find Top 5 Most common Jobs: 
data["JobTitle"].value_counts().head(5)

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64